### ISSUE TITLE
[MNG-6164] Collections inconsistently immutable

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7280]
**[Christian Schulte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=schulte77)** opened **[MNG-6164](https://issues.apache.org/jira/browse/MNG-6164?redirect=false)** and commented

There are plenty of places where empty collections are returned from public API in methods written like:

```
public List<Exception> getExceptions()
{
   return exceptions == null ? Collections.<Exception>emptyList() : exceptions;
}
```

The issue with this is that the empty list is immutable but the collection returned for the nun-null case is not immutable.

All those methods should return a collection with consistent "mutability": either mutable, either immutable.

Given empty immutable collections do not cause harm until now, switching consistently to immutable collections is more conservative and should not be risky

---

**Affects:** 3.5.0

**Remote Links:**
- [GitHub Pull Request #141
](https://github.com/apache/maven/pull/141)
- [GitHub Pull Request #141
](https://github.com/apache/maven/pull/141)

0 votes, 6 watchers

**Comment by jira-importer** [comment]
**[Christian Schulte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=schulte77)** commented

Issue is fixed in the master branch of [my private maven repository on github](https://github.com/ChristianSchulte/maven/tree/master). When interested, you can cherry pick the commit to your own repository and create a pull request for the apache master branch with just this commit yourself from there.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise), do you want to reassign this ticket to do since you did the commit?

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

Hi [Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o), I just rebased that branch...but of course I can reassign if you don't mind?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise), can we make this part of 3.6.0 finally?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [44826ab446d1115d464e73e7e308df36dcf7d39b](https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=44826ab446d1115d464e73e7e308df36dcf7d39b).
